### ISSUE TITLE
rec: address the aggresive cache pruning issues mentioned in #13106

### ISSUE THREAD

**Issue by omoerbeek** [PowerDNS/pdns#13109]
In some cases) pruning is unfair (it scans the zones always in the same order and stops clearing when it has reached the goal) and/or not very active (when the recursor is lightly loaded). this has the consequence that old expired records can remain in the cache that prevent new data to be recorded and used (unless #13106 is merged).

On a lightly loaded rec I see this with `rec_control dump-cache -`:
```
; Zone .
. -36987 IN NSEC aaa. NS SOA RRSIG NSEC DNSKEY
```
Note that the root zone is last in the list, so entries in it are most likely to be skipped when pruning without pressure.

The unfairness should be fixed, possibly using recent changes in ``cachecleaner.hh`` as a guide.

There's also another issue with pruning, I see on another rec in the aggressive cache:
```
...
; Zone nl.
; Zone org.
; Zone pro.
; Zone pub.
; Zone tv.
; Zone co.uk.
; Zone video.
; Zone world.
; Zone .
...
```
from looking at the pruning code these empty zones should have been cleaned up. Maybe they are created empty?

**Comment by omoerbeek** [comment]
The current pruning code uses two approaches. For each approach the zones are scannes in the same order, and for each zone entries are scanned in LRU order, potentially removing some.

1. if the cache more entries than the max, it will scan `toLook` entries and remove expired ones until it has removed `toErase` entries. The issue here is that both counts are global count for the whole cache. So if the first few zones scanned already produce enough removals, the latter zones will not be scanned and will not be pruned.
2. If the cache is not full, a different `toLook` is computed, and the zones are scanned. For each zone, entries are scanned and erase if they are expired, until `toLook` entries are examined.

So to summarize, I can see several problems with this approach:
- The zones at the end are not scanned at all if the goal is reached earlier
- The number of entries per zone can vary wildly. I think you want to scan a fraction of the entries depending on the size of the zone. This is what the code in `cachecleaner.hh` does.
- Even if the cache is full, only *expired* entries are removed. This gives no guarantee that actual trimming takes place.
- A specific issue is the `if (it->d_ttd < now || lookedAt > toLook) { // erase }` statement in the "cache not full" case.  I don't understand the rhs of the condition.
